test/unit/test-controller-promise.js

```diff
@@ -0,0 +1,278 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ControllerPromise} from '../../build-system/tasks/e2e/controller-promise';
+
+// Since this is designed to run in Node we don't need to run it
+// on the full set of sauce labs browsers
+const config = describe.configure().ifChrome();
+config.run('ControllerPromise', () => {
+  describe('Promise wrapping behavior', () => {
+    it('should behave like a normal thenable', () => {
+      const p = new ControllerPromise(Promise.resolve('success'));
+      return p.then(result => expect(result).to.equal('success'));
+    });
+
+    it('should behave like a normal thenable with await', async () => {
+      const p = new ControllerPromise(Promise.resolve('success'));
+      await expect(await p).to.equal('success');
+    });
+
+    it('should accept calls to `then`', async () => {
+      const p = new ControllerPromise(Promise.resolve(1));
+      const two = p.then(x => x + 1);
+
+      const three = two.then(x => x + 1);
+      const four = two.then(x => x + 2);
+
+      await expect(await two).to.equal(2);
+      await expect(await three).to.equal(3);
+      await expect(await four).to.equal(4);
+    });
+
+    it('should allow empty `then` calls', async () => {
+      const p = new ControllerPromise(Promise.resolve(1));
+      const two = p.then(null).then(x => x + 1);
+
+      const three = two.then(x => x + 1);
+      const four = two.then(x => x + 2);
+
+      await expect(await two).to.equal(2);
+      await expect(await three).to.equal(3);
+      await expect(await four).to.equal(4);
+    });
+
+    it('should accept calls to `catch`', async () => {
+      const p = new ControllerPromise(Promise.reject('failure'));
+      p.then(() => {
+        throw new Error('should not happen');
+      });
+      return p.catch(e => {
+        expect(e).to.equal('failure');
+      });
+    });
+
+    it('should accept a second parameter for `then`', async () => {
+      const p = new ControllerPromise(Promise.reject('failure'));
+      p.then(() => {
+        throw new Error('should not happen');
+      });
+      return p.then(null, e => {
+        expect(e).to.equal('failure');
+      });
+    });
+
+    it('should accept calls to `finally`', () => {
```

Comment:
Would add a test for `finally` semantics, for example that it passes through errors. That is, something like:

```javascript
const finallySpy = sinon.spy();
const catchSpy = sinon.spy();
const failureObject = {};
const rejectedControllerPromise = new ControllerPromise(
  Promise.reject(failureObject)
);
await rejectedControllerPromise
  .finally(finallySpy)
  .catch(catchSpy);

expect(finallySpy).to.have.been.calledOnce;
expect(catchSpy).to.have.been.calledWith(failureObject);
```

Reply:
Sounds good. Using spies makes a lot of sense too, I'll try that in some of the other tests as well